HW2/P2/parallel_vector.pyx

```diff
@@ -99,17 +99,17 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
        int num_locks = (counts.shape[0] + N - 1) / N  # ensure enough locks
        omp_lock_t *locks = get_N_locks(num_locks)
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for every N adjacent elements of counts
-   # to parallelize data movement.  Be sure to avoid deadlock, as well as
-   # double-locking.
-   ##########
    with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
+      for r in range(repeat):
+         for idx in prange(src.shape[0], num_threads = 4):
+            if src[idx] != dest[idx]: 
                if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+                  acquire(&(locks[dest[idx] / N]))
+                  counts[dest[idx]] += 1
+                  release(&(locks[dest[idx] / N]))
+
+                  acquire(&(locks[src[idx] / N])) 
+                  counts[src[idx]] -= 1
+                  release(&(locks[src[idx] / N]))    
```

Comment:
No. Some stuff came up and I ended up not being able to finish this pset at all.